core/src/main/java/org/apache/calcite/plan/volcano/VolcanoRuleMatchVisualizer.java

```diff
@@ -0,0 +1,312 @@
+    }
+
+    this.addRuleMatch("FINAL", new ArrayList<>(finalPlanNodes));
+  }
+
+  private String getSetLabel(RelSet set) {
+    return "set-" + set.id + "    ";
+  }
+
+  private String getJsonStringResult() {
+    try {
+      RelNode root = volcanoPlanner.getRoot();
+      if (root == null) {
+        throw new RuntimeException("volcano planner root is null");
+      }
+      RelMetadataQuery mq = root.getCluster().getMetadataQuery();
+
+      Map<String, VisualizerNodeInfo> nodeInfoMap = new TreeMap<>();
+      for (Map.Entry<String, RelNode> entry : allNodes.entrySet()) {
+        RelNode relNode = entry.getValue();
+        RelOptCost cost = volcanoPlanner.getCost(relNode, mq);
+        Double rowCount = mq.getRowCount(relNode);
+
+        VisualizerNodeInfo nodeInfo;
+        if (relNode instanceof RelSubset) {
+          RelSubset relSubset = (RelSubset) relNode;
+          String nodeLabel = "subset#" + relSubset.getId() + "-set#" + relSubset.set.id + "-\n"
+              + relSubset.getTraitSet().toString();
+          String relIDs = relSubset.getRelList().stream()
+              .map(i -> "#" + i.getId()).collect(joining(", "));
+          String explanation = "rels: [" + relIDs + "]";
+          nodeInfo =
+              new VisualizerNodeInfo(nodeLabel, true, explanation, cost, rowCount);
+        } else {
+          InputExcludedRelWriter relWriter = new InputExcludedRelWriter();
+          relNode.explain(relWriter);
+          String inputIDs = relNode.getInputs().stream()
+              .map(i -> "#" + i.getId()).collect(joining(", "));
+          String explanation = relWriter.toString() + ", inputs: [" + inputIDs + "]";
+
+          String nodeLabel = "#" + relNode.getId() + "-" + relNode.getRelTypeName();
+          nodeInfo = new VisualizerNodeInfo(nodeLabel, false, explanation, cost,
+              rowCount);
+        }
+
+        nodeInfoMap.put(entry.getKey(), nodeInfo);
+      }
+
+      LinkedHashMap<String, Object> data = new LinkedHashMap<>();
+      data.put("allNodes", nodeInfoMap);
+      data.put("ruleMatchSequence", ruleMatchSequence);
+      data.put("ruleMatchInfoMap", ruleInfoMap);
+      data.put("nodeAddedInRule", nodeAddedInRule);
+
+      ObjectMapper objectMapper = new ObjectMapper();
+      return objectMapper.writerWithDefaultPrettyPrinter().writeValueAsString(data);
+    } catch (JsonProcessingException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  /**
+   * Writes the HTML and JS files of the rule match visualization.
+   * <p>
+   * The old files with the same name will be replaced.
+   *
+   * @param outputDirectory directory of the output files
+   * @param suffix          file name suffix, can be null
+   */
+  public void writeToFile(String outputDirectory, String suffix) {
+    // default HTML template is under "resources"
+    writeToFile("volcano-viz", outputDirectory, suffix);
+  }
+
+  public void writeToFile(String templateDirectory, String outputDirectory, String suffix) {
```

Comment:
Instead of forcing the client to keep a reference to the visualizer and call these methods explicitly after optimization can't we do it automatically at the end of the optimization phase (e.g., by relying on `RelOptListener#relChosen`)? 
If this gets integrated to the planner we should find a trick to avoid replacing the visualizations of previous queries; in other words avoid writing to the same file/directory.

Reply:
I've changed it so that the method will be called automatically when using a `VolcanoPlaner`. It has to be called manually for a `HepPlanner`. The directory and the suffix will have to be chosen when creating the visualizer.